### ISSUE TITLE
Leverage `clonefile(2)` in `naive` implementation when in Darwin on top of APFS-based mountpoint

### ISSUE THREAD

**Issue by cirocosta** [concourse/concourse#3518]
Hey,

I noticed that under our [Baggageclaim](https://github.com/concourse/baggageclaim) `naive`  driver implementation use a plain recursive `cp`:

```go
func (driver *NaiveDriver) CreateCopyOnWriteLayer(path string, parent string) error {
	return exec.Command("cp", "-rp", parent, path).Run()
}
```
(*https://github.com/concourse/baggageclaim/blob/8f1eaeab1832501f1488a8b966bbf80f4307d0ad/volume/driver/naive_stub.go#L9-L11*)

But, we could possibly use `-c` in Darwin when using APFS to take advantage of the `clonefile(2)` capabilities:

```
man cp
    ...
     -c    copy files using clonefile(2)
``` 

```
man clonefile
NAME
     clonefile -- create copy on write clones of files

DESCRIPTION
     The clonefile() function causes the named file src to be cloned to the named file dst.  The cloned file dst shares its data blocks with the src file but has its own copy of attributes, extended attributes and
     ACL's which are identical to those of the named file src with the exceptions listed below
...
```

See https://developer.apple.com/documentation/foundation/file_system/about_apple_file_system

Here are some stats from `cp`ing things around:

1. populate 1000 files with 10M each

```sh
seq 1 1000 | xargs -I[] -P 4 dd if=/dev/zero of=./[] bs=1024 count=10240
```

2. copy without `-c`
 
```sh
 ~/test $ time cp -rp ./a ./b

real	0m12.866s
user	0m0.069s
sys	0m6.819s
```

3. copy with `-c`

```sh
 ~/test $ time cp -crp ./a ./c

real	0m0.142s
user	0m0.005s
sys	0m0.109s
```

**Comment by cirocosta** [comment]
cc @ddadlani @kcmannem

**Comment by marco-m** [comment]
@cirocosta good catch! This might explain some slowness we see in our Mac workers when moving volumes around.

**Comment by taylorsilva** [comment]
# Issue Bankruptcy

This issue is being closed as part of 2025's **Issue Bankruptcy** declaration. [See here for details](https://blog.concourse-ci.org/posts/2025-01-06-concourse-2024-review/)

If you think this issue still has merit then feel free to open a new issue and reference this one.

---

This looks like it could be easy to fix, so anyone is free to PR a fix to this. Baggageclaim is in the main repo now: https://github.com/concourse/concourse/tree/master/worker/baggageclaim
